Add a flag to use FusedMoE kernel in compressed quantization

chenxi-yang opened the pull request:
Summary: Allows customizing kernel usage in compressed-tensor quantization.

Test Plan:
CUDA_VISIBLE_DEVICES=6,7 \
  VLLM_DISABLE_COMPILE_CACHE=1 \
  VLLM_MQ_MAX_CHUNK_BYTES_MB=256 \
  VLLM_GPU_MEMORY_UTILIZATION=0.85 \
  buck2 run @//mode/{opt,inplace} \
    -c fbcode.enable_vllm=true \
    -c fbcode.enable_gpu_sections=true \
    -c fbcode.nvcc_arch=h100a \
    //smart/inference_platform_sp/llm_predictor_gpu:service -- \
    --local_cache_dir "$HOME/local/models/GLM-4.5V-FP8" \
    --try_local_cache \
    --max_seq_len=16384 \
    --max_batch_size 192 \
    --thrift_server_port 12345 \
    --enable_warmup=true \
    --model_mf_bucket=llm_inference \
    --model_mf_path=tree/oss/GLM-4.5V-FP8 \
    --force_llm_format=true \
    --allow_custom_stop_tokens \
    --model_parallel_size 2 \
    --vllm_engine \
    --cpu_offload_gb=0 \
    --kv_cache_quantization 8

Before:

QPS:                 1.26
Avg latency:         49.998s
Avg TTFT (client):   1679.44ms
P50 TTFT (client):   1584.17ms
P99 TTFT (client):   5748.46ms
Avg TTIT (client):   48.32ms
P50 TTIT (client):   48.21ms
P99 TTIT (client):   59.81ms
Avg TTFT (server):   2481.96ms
Avg TTIT (server):   48.06ms
Avg prefill len:     2643.00 tokens
P50 prefill len:     2643.00 tokens
P99 prefill len:     2643.00 tokens
Avg decode len:      1000.00 tokens
P50 decode len:      1000.00 tokens
P99 decode len:      1000.00 tokens

After:

QPS:                 1.86
Avg latency:         92.238s
Avg TTFT (client):   1856.53ms
P50 TTFT (client):   1912.39ms
P99 TTFT (client):   2694.97ms
Avg TTIT (client):   70.38ms
P50 TTIT (client):   75.11ms
P99 TTIT (client):   76.74ms
Avg TTFT (server):   2984.87ms
Avg TTIT (server):   77.98ms
Avg prefill len:     2643.00 tokens
P50 prefill len:     2643.00 tokens
P99 prefill len:     2643.00 tokens
Avg decode len:      1000.00 tokens
P50 decode len:      1000.00 tokens
P99 decode len:      1000.00 tokens

Rollback Plan:

Differential Revision: D80552023

yewentao256 reviewed (commented):
Wondering why do we need this flag, could you give more context?

chenxi-yang commented on vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py:
```diff
@@ -847,8 +848,9 @@ def apply(
-            # small-batch fallback on SM100
-            if self.is_fp8_w8a8_sm100 and topk_ids.shape[0] <= 8:
+            # SM90 or small-batch fallback on SM100
+            if self.is_fp8_w8a8_sm90 or (
```
Thanks for the explanation!
For now, I am planning to test the following for `glm-4.5v-fp8` and `Llama-4-Maverick-17B-128E-Instruct-FP8`:
1. `lm_eval` both models' accuracy with and without this PR. (the maverick-fp8 results are already shown above)
2. benchmark the `fused_moe() [with optimal kernel config]` and `cutlass_moe_fp8()` for `glm-4.5v-fp8` and `Llama-4-Maverick-17B-128E-Instruct-FP8`.
https://github.com/vllm-project/vllm/blob/1c3dad22ff92cbf84e0fa8ad1643c560a07944ea/vllm/model_executor/layers/fused_moe/fused_moe.py#L1211

https://github.com/vllm-project/vllm/blob/1c3dad22ff92cbf84e0fa8ad1643c560a07944ea/vllm/model_executor/layers/fused_moe/cutlass_moe.py#L392

What do you think? @yewentao256 @houseroad

chenxi-yang replied:
@yewentao256 @houseroad I added the cutlass moe fp8 benchmark here: https://github.com/vllm-project/vllm/pull/25302, PTAL. glm config is here: https://github.com/vllm-project/vllm/pull/24911

Triton fused moe with optimal config is generally faster by > 20% than cutlass moe. However, triton fused is worse than cutlass when using default config. The gain seems to come from a combination of fusion and kernel tuning. As long as one tunes the kernel config, it should be good to go with fused moe for now.

In the future, there may be some optimization options: 1). add fused cutlass moe 2). customize vllm moe options

Here are the details:

#### glm4.5v-fp8 setting

##### triton with op config
<img width="576" height="462" alt="image" src="https://github.com/user-attachments/assets/5658a6b3-d3f5-4c34-b8a6-24c6907010bf" />

##### triton with default config
<img width="602" height="455" alt="image" src="https://github.com/user-attachments/assets/ce82f023-bf8c-42ca-a05b-08787c563891" />

#### Llama-4-Maverick-17B-128E-Instruct-FP8 setting

##### triton with op config
<img width="637" height="469" alt="image" src="https://github.com/user-attachments/assets/1f58f895-db56-4a16-b40c-9d986ecccf8d" />

##### triton with default config
<img width="560" height="421" alt="image" src="https://github.com/user-attachments/assets/c92a6a9f-7487-4166-8f7b-cfabcc8bbfc5" />

yewentao256 replied:
>lm_eval both models' accuracy with and without this PR. (the maverick-fp8 results are already shown above)

This looks good to me, could you add report for that?

chenxi-yang committed:
Add a flag to use FusedMoE kernel in compressed quantizationon on sm90 (#23442)

Summary:
Pull Request resolved: https://github.com/vllm-project/vllm/pull/23442

Allows fused moe kernel usage in compressed-tensor quantization on sm90.

Signed-off-by: Chenxi Yang <cxyang@meta.com>

Test Plan:
CUDA_VISIBLE_DEVICES=6,7 \
  VLLM_DISABLE_COMPILE_CACHE=1 \
  VLLM_MQ_MAX_CHUNK_BYTES_MB=256 \
  VLLM_GPU_MEMORY_UTILIZATION=0.85 \
  buck2 run @//mode/{opt,inplace} \
    -c fbcode.enable_vllm=true \
    -c fbcode.enable_gpu_sections=true \
    -c fbcode.nvcc_arch=h100a \
    //smart/inference_platform_sp/llm_predictor_gpu:service -- \
    --local_cache_dir "$HOME/local/models/GLM-4.5V-FP8" \
    --try_local_cache \
    --max_seq_len=16384 \
    --max_batch_size 192 \
    --thrift_server_port 12345 \
    --enable_warmup=true \
    --model_mf_bucket=llm_inference \
    --model_mf_path=tree/oss/GLM-4.5V-FP8 \
    --force_llm_format=true \
    --allow_custom_stop_tokens \
    --model_parallel_size 2 \
    --vllm_engine \
    --cpu_offload_gb=0 \
    --kv_cache_quantization 8

Before:

QPS:                 1.26
Avg latency:         49.998s
Avg TTFT (client):   1679.44ms
P50 TTFT (client):   1584.17ms
P99 TTFT (client):   5748.46ms
Avg TTIT (client):   48.32ms
P50 TTIT (client):   48.21ms
P99 TTIT (client):   59.81ms
Avg TTFT (server):   2481.96ms
Avg TTIT (server):   48.06ms
Avg prefill len:     2643.00 tokens
P50 prefill len:     2643.00 tokens
P99 prefill len:     2643.00 tokens
Avg decode len:      1000.00 tokens
P50 decode len:      1000.00 tokens
P99 decode len:      1000.00 tokens

After:

QPS:                 1.86
Avg latency:         92.238s
Avg TTFT (client):   1856.53ms
P50 TTFT (client):   1912.39ms
P99 TTFT (client):   2694.97ms
Avg TTIT (client):   70.38ms
P50 TTIT (client):   75.11ms
P99 TTIT (client):   76.74ms
Avg TTFT (server):   2984.87ms
Avg TTIT (server):   77.98ms
Avg prefill len:     2643.00 tokens
P50 prefill len:     2643.00 tokens
P99 prefill len:     2643.00 tokens
Avg decode len:      1000.00 tokens
P50 decode len:      1000.00 tokens
P99 decode len:      1000.00 tokens

The accuracy is 0.74 (w/ this diff and w/o this diff are on par)

Reviewed By: zzh142857, wangwenchen0407

Differential Revision: D80552023

chenxi-yang committed:
Merge branch 'main' into export-D80552023
